File: android/app/src/main/java/org/maturita/maturita/data/Json.kt
package org.maturita.maturita.data

import org.json.JSONArray
import org.json.JSONObject

class J(val o: JSONObject) {
    fun has(key: String) = o.has(key) && !o.isNull(key)
    fun str(key: String, fallback: String = "") =
        if (!has(key)) fallback else o.optString(key, fallback)
    fun strOrNull(key: String): String? =
        if (!has(key)) null else o.optString(key).takeIf { it.isNotEmpty() && it != "null" }
    fun bool(key: String, fallback: Boolean = false) = o.optBoolean(key, fallback)
    fun int(key: String, fallback: Int = 0) = o.optInt(key, fallback)
    fun obj(key: String): J? = o.optJSONObject(key)?.let { J(it) }
    fun exercise(n: Int): J? = o.optJSONObject("exercises")?.optJSONObject(n.toString())?.let { J(it) }
    fun arr(key: String): List<J> {
        val a = o.optJSONArray(key) ?: return emptyList()
        return (0 until a.length()).mapNotNull { i -> a.optJSONObject(i)?.let { J(it) } }
    }
    fun strs(key: String): List<String> {
        val a = o.optJSONArray(key) ?: return emptyList()
        return (0 until a.length()).mapNotNull { i ->
            if (a.isNull(i)) null else a.optString(i).takeIf { it.isNotEmpty() }
        }
    }
    fun strsOrEmpty(key: String): List<String?> {
        val a = o.optJSONArray(key) ?: return emptyList()
        return (0 until a.length()).map { i ->
            if (a.isNull(i)) null else a.optString(i)
        }
    }
    fun intRows(key: String): List<List<String>> {
        val a = o.optJSONArray(key) ?: return emptyList()
        return (0 until a.length()).map { i ->
            val row = a.optJSONArray(i) ?: JSONArray()
            (0 until row.length()).map { j -> row.optString(j) }
        }
    }

    fun answerList(): List<String> {
        val a = o.optJSONArray("answers")
        if (a != null) {
            return (0 until a.length()).mapNotNull { i ->
                if (a.isNull(i)) null else a.optString(i).takeIf { it.isNotEmpty() }
            }
        }
        return listOfNotNull(strOrNull("answers"))
    }
}

fun JSONArray.objects(): List<J> =
    (0 until length()).mapNotNull { i -> optJSONObject(i)?.let { J(it) } }

File: ios/Maturita/Data/JSON.swift
import Foundation

struct J {
    let o: [String: Any]

    init(_ obj: Any?) {
        o = obj as? [String: Any] ?? [:]
    }

    func has(_ key: String) -> Bool {
        guard let value = o[key], !(value is NSNull) else { return false }
        return true
    }

    func str(_ key: String, _ fallback: String = "") -> String {
        guard has(key) else { return fallback }
        if let s = o[key] as? String { return s }
        if let n = o[key] as? NSNumber { return n.stringValue }
        return fallback
    }

    func strOrNull(_ key: String) -> String? {
        guard has(key) else { return nil }
        if let s = o[key] as? String, !s.isEmpty, s != "null" { return s }
        return nil
    }

    func bool(_ key: String, _ fallback: Bool = false) -> Bool {
        if let b = o[key] as? Bool { return b }
        if let n = o[key] as? NSNumber { return n.boolValue }
        return fallback
    }

    func int(_ key: String, _ fallback: Int = 0) -> Int {
        if let n = o[key] as? Int { return n }
        if let n = o[key] as? NSNumber { return n.intValue }
        if let s = o[key] as? String, let n = Int(s) { return n }
        return fallback
    }

    func obj(_ key: String) -> J? {
        guard let value = o[key], !(value is NSNull) else { return nil }
        if value is [String: Any] { return J(value) }
        return nil
    }

    func exercise(_ n: Int) -> J? {
        guard let ex = o["exercises"] as? [String: Any] else { return nil }
        return J(ex["\(n)"])
    }

    func arr(_ key: String) -> [J] {
        guard let a = o[key] as? [Any] else { return [] }
        return a.compactMap { item in
            item is [String: Any] ? J(item) : nil
        }
    }

    func strs(_ key: String) -> [String] {
        guard let a = o[key] as? [Any] else { return [] }
        return a.compactMap { item in
            if item is NSNull { return nil }
            if let s = item as? String, !s.isEmpty { return s }
            return nil
        }
    }

    func strsOrEmpty(_ key: String) -> [String?] {
        guard let a = o[key] as? [Any] else { return [] }
        return a.map { item in
            if item is NSNull { return nil }
            return item as? String
        }
    }

    func intRows(_ key: String) -> [[String]] {
        guard let a = o[key] as? [Any] else { return [] }
        return a.map { row in
            guard let cells = row as? [Any] else { return [] }
            return cells.map { cell in
                if let s = cell as? String { return s }
                if let n = cell as? NSNumber { return n.stringValue }
                return ""
            }
        }
    }

    func rawArray(_ key: String) -> [Any] {
        o[key] as? [Any] ?? []
    }

    func answerList() -> [String] {
        if let a = o["answers"] as? [Any] {
            return a.compactMap { item in
                if item is NSNull { return nil }
                if let s = item as? String, !s.isEmpty { return s }
                return nil
            }
        }
        return [strOrNull("answers")].compactMap { $0 }
    }
}
